text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+
+# Design Summary
+
+In this proposal we introduce two new constructs:
+
+- `Environment` - A simple construct which wraps up a VPC, cluster, and other top level
+  resources needed for ECS services to run on an AWS account
+- `Service` - A single application that runs inside an `Environment`
+
+And two new associated foundational classes used by the `Service` construct:
+
+- `ServiceDescription` - A class which is utilized to build out a description of the
+  application and the features that should be attached to the application. This description
+  is consumed by the `Service` class
+- `ServiceExtension` - An abstract class for a single type of feature that you want to
+  attach to a service. A `ServiceExtension` is added to a `ServiceDescription` and then
+  the `ServiceDescription` is used to make a new `Service` construct.
+
+The following code example summarizes how you would use these classes together:
+
+```js
+const myServiceDesc = new ServiceDescription();
+myServiceDesc.add(new Container({
+  cpu: 1024,
+  memoryMiB: 2048,
+  trafficPort: 80,
+  image: ContainerImage.fromRegistry('nathanpeck/greeter')
+}));
+myServiceDesc.add(new AppMeshExtension({ mesh }));
+myServiceDesc.add(new FireLensExtension());
+myServiceDesc.add(new XRayExtension());
+myServiceDesc.add(new CloudwatchAgentExtension());
+myServiceDesc.add(new HttpLoadBalancerExtension());
+
+const myEnvironment = new Environment(stack, 'production', {
+  vpc: vpc,
+  cluster: cluster,
+  capacityType: capacityTypes.EC2 || capacityTypes.FARGATE
+});
+
+const myService = new Service(stack, 'my-service', {
+  environment: myEnvironment
+  serviceDescription: myServiceDesc
+});
+```
+
+# Detailed Design
+
+Let's take a more detailed look at each class:
+
+## Class: `ServiceDescription`
+
+This class is not an actual construct, it is just a helper class which is used to build
+out a description of the service to build. This class is used because traditionally all
+CDK constructs are immutable. Once a `Service` is created using its constructor most if
+not all of its properties should be immutable. Therefore its necessary to build out a
+full list of all the things that need to be constructed, prior to construction time.
+
+The `ServiceDescription` class has one method: `ServiceDescription.add(ServiceExtension)`.
+Usage looks similar to this:
+
+```js
+const myServiceDesc = new ServiceDescription();
+myServiceDesc.add(new Container();
+myServiceDesc.add(new AppMeshExtension({ mesh }));
+myServiceDesc.add(new FireLensExtension());
+myServiceDesc.add(new XRayExtension());
+```
+
+## Class: `ServiceExtension`
+
+This class defines an actual extension itself. The extension is collection of hooks which
+modify the `Service` properties, `TaskDefinition` properties, and/or make use of the
+resulting `Service` and `TaskDefinition`. Each `ServiceExtension` can implement the
+following hooks:
+
+- `addHooks()` - This hook is called after all the extensions are added to a
```

Comment:
I would try to keep this API as small as possible. IF (big if) the constructs were mutable, then the `useXxx()` APIs would mostly be all that's required, correct?

Reply:
Unfortunately many of the construct properties are not mutable. Questionable whether we want to make the L2's fully mutable or use a multi step process of generating the props first then creating?